collectors/charts.d.plugin/charts.d.plugin.in

```diff
@@ -612,6 +612,12 @@ debug "run_charts='$run_charts'"
 
 [ -z "$run_charts" ] && fatal "No charts to collect data from."
 
+keepalive() {
+  if [ ! -t 1 ] && ! printf "\n"; then
+    chartsd_cleanup
+  fi
+}
```

Comment:
So if I am to understand this piece of code...

> While we are trying to attempt an upgrade; If we are not running inside a terminal and our `write()` syscall fails to write to FD `1` then we assume we still have a hung process and perform `chartsd_cleanup`.

Is this right?

Reply:
Yes, `\n` is the only thing we can write as a keepalive message according external plugins API. If pipe is broken `printf` returns error and we need to exit. That is the only way to not have zombie process on old systems. For some reason Netdata process doesnt send sigterm to its children or something.

Reply:
`[ ! -t 1 ]` part is completely optional, but we dont need the check on terminal

Reply:
Thia now all makes sense to me. Thank you! The fact that the NetData agent doesn't correctly send `SIGTERM` to all its child processes _may_ be a bug maybe or maybe we should consider using process groups. cc @manos-saratsis

Reply:
i am not sure it is a bug, i was able to reproduce only with following conditions:
 - send `SIGKILL` to netdata
 - os: centos 6


There is no such problem on deb9, arch for example.